### ISSUE TITLE
-Woverloaded-shift-op-parentheses does not detect issues with implicit cast to bool

### ISSUE THREAD

**Issue by PiotrZSL** [llvm/llvm-project#61943]
Example:

This one reports `-Woverloaded-shift-op-parentheses`:
```c++
#include <iostream>

int main() {
    int a = 5;
    std::cout << "Test" << a == 5 ? 1 : 0;
    return 0;
}
```
```console
warning: overloaded operator << has higher precedence than comparison operator [-Woverloaded-shift-op-parentheses]
```

But this one not:
```c++
#include <iostream>

int main()
{
    int a = 5;
    std::cout << "Test" << a ? 1 : 0;
    return 0;
}
```

And this one:
```c++
#include <iostream>

int main()
{
    int a = 5;
    std::cout << "Test" << static_cast<bool>(a) ? 1 : 0;
    return 0;
}
```
reports:
```console
warning: operator '?:' has lower precedence than '<<'; '<<' will be evaluated first [-Wparentheses]
```

I would expect that all of above report some sort of warning.
This is reported by Sonar as CWE-783:  https://cwe.mitre.org/data/definitions/783.html
> Suspicious calculation. Please use parentheses to clarify the code. The code ''a+b?c:d'' should be written as either ''(a+b)?c:d'' or ''a+(b?c:d)''.

This should be handled by some warning, or by some Clang-Tidy check.

**Comment by shafik** [comment]
So it looks like in `DiagnoseBinOpPrecedence(...)`:

```cpp
if (BinaryOperator::isComparisonOp(Opc))
    DiagnoseShiftCompare(Self, OpLoc, LHSExpr, RHSExpr);
```

`isComparisonOp(...)`  is limiting to these cases `return Opc >= BO_Cmp && Opc<=BO_NE;` 

So I think this should be expandable to more cases and maybe even a good fist problem wdyt @AaronBallman

**Comment by AaronBallman** [comment]
I think it makes sense to add diagnostics to these cases, but we should validate my assumption that doing so will predominately catch true positives. As for whether this is a good first issue or not... it seems reasonable to me.

**Comment by chaitanyav** [comment]
@AaronBallman @shafik can i look into this one?

**Comment by AaronBallman** [comment]
> @AaronBallman @shafik can i look into this one?

Absolutely! (FWIW, anyone can work on any issued they'd like -- no permission needed.)

**Comment by krishenm94** [comment]
hey @chaitanyav , are you still working on this. Could I pick it up?

**Comment by chaitanyav** [comment]
@krishenm94 am working on it, want to get some feedback from @AaronBallman on this https://reviews.llvm.org/D147844

**Comment by chaitanyav** [comment]
@krishenm94 please go ahead. I will assign this to you.

**Comment by AaronBallman** [comment]
@chaitanyav -- sorry for the delay in getting you that feedback; FWIW, it sometimes takes a week or two for reviewers to review things, depending on how heavy our review backlog is/vacations/other responsibilities.

FWIW, your patch was basically ready. I spotted one question in the test cases (is the change to `operator bool()` needed for anything?) and that it's missing a release note, but that's basically it.

**Comment by chaitanyav** [comment]
@AaronBallman let me send out another revision with the changes.

**Comment by Endilll** [comment]
@chaitanyav are you still working on this?

**Comment by chaitanyav** [comment]
https://reviews.llvm.org/D147844, it was decided it is more of a style issue. i think no more work should be done on this.

**Comment by erichkeane** [comment]
See discussion on D147844
